transport/httptransport/libp2p_server.go

```diff
@@ -0,0 +1,372 @@
+		for l := range s.eventListeners {
+			(*l)(val.DBID, evt)
+		}
+		s.eventListenersLk.Unlock()
+	}
+
+	s.transfersLk.Lock()
+	existing, ok := s.transfers[val.ProposalCid]
+	if ok {
+		log.Infow("restarting transfer", "proposalCID", val.ProposalCid, "payloadCID", val.PayloadCid)
+		existing.cancelTimeout()
+	} else {
+		log.Infof("starting transfer", "proposalCID", val.ProposalCid, "payloadCID", val.PayloadCid)
+	}
+	xfer := &transferTimer{
+		authValue:    &val,
+		retryTimeout: s.cfg.RetryTimeout,
+		onTimeout: func(orig error) {
+			s.transfersLk.Lock()
+			delete(s.transfers, val.ProposalCid)
+			s.transfersLk.Unlock()
+
+			log.Infof("failed transfer after timeout", "proposalCID", val.ProposalCid, "payloadCID", val.PayloadCid,
+				"error", orig, "timeout", s.cfg.RetryTimeout)
+			_ = s.Cleanup(authToken)
+
+			msg := fmt.Sprintf("timed out waiting %s for data receiver to restart transfer ", s.cfg.RetryTimeout)
+			msg += "after error " + orig.Error()
+			msg += "\nfor proposal CID " + val.ProposalCid.String() + ", payloadCID " + val.PayloadCid.String()
+			fireEvent(transferred, types.TransferStatusFailed, msg)
+		},
+	}
+	s.transfers[val.ProposalCid] = xfer
+	s.transfersLk.Unlock()
+
+	// Fire event before starting transfer
+	fireEvent(0, types.TransferStatusStarted, "")
+
+	// Fire progress events during transfer.
+	// We fire a progress event each time bytes are read from the CAR.
+	// Note that we can't fire events when bytes are written to the HTTP stream
+	// because there may be some transformation first (eg adding headers etc).
+	readEmitter := &readEmitter{ReadSeeker: content, emit: func(count int, err error) {
+		if err != nil {
+			xfer.startRetryTimer(err)
+			return
+		}
+		transferred += uint64(count)
+		fireEvent(transferred, types.TransferStatusOngoing, "")
+	}}
+
+	// http.ServeContent ignores errors when writing to the stream, so we
+	// replace the writer with a class that watches for errors
+	writeErrWatcher := &writeErrorWatcher{ResponseWriter: w, onError: xfer.startRetryTimer}
+
+	http.ServeContent(writeErrWatcher, r, "", time.Time{}, readEmitter)
+
+	// Check if there was an error during the transfer
+	if xfer.errored() {
+		// There was an error, just wait for the data receiver to retry
+		return
+	}
+
+	s.transfersLk.Lock()
+	delete(s.transfers, val.ProposalCid)
+	s.transfersLk.Unlock()
+
+	// Fire event at the end of the transfer
+	fireEvent(transferred, types.TransferStatusCompleted, "")
+
+	log.Infow("completed serving request", "proposalCID", val.ProposalCid, "payloadCID", val.PayloadCid)
+}
+
+type EventListenerFn func(dbid uint, st types.TransferState)
+type UnsubFn func()
+
+func (s *Libp2pCarServer) Subscribe(cb EventListenerFn) UnsubFn {
+	s.eventListenersLk.Lock()
+	s.eventListeners[&cb] = struct{}{}
+	s.eventListenersLk.Unlock()
+
+	return func() {
+		s.eventListenersLk.Lock()
+		delete(s.eventListeners, &cb)
+		s.eventListenersLk.Unlock()
+	}
+}
+
+type transferTimer struct {
```

Comment:
Let's remove this completely and move this to a higher layer as discussed offline.